lib/client/conntest/ssh.go

```diff
@@ -231,10 +239,45 @@ func (s SSHConnectionTester) handleErrFromSSH(ctx context.Context, connectionDia
 	}
 
 	processStdoutString := strings.TrimSpace(processStdout.String())
-	if strings.HasPrefix(processStdoutString, "Failed to launch: user:") {
+	// If the selected principal does not exist on the node, attempting to connect emits:
+	// "Failed to launch: user: lookup username <principal>: no such file or directory."
+	isUsernameLookupFail := strings.HasPrefix(processStdoutString, "Failed to launch: user:")
+	// Connect My Computer runs the agent as non-root. When attempting to connect as another system
+	// user that is not the same as the user who runs the agent, the emitted error is "Failed to
+	// launch: fork/exec <conn.User shell>: operation not permitted."
+	isForkExecOperationNotPermitted := strings.HasPrefix(processStdoutString, "Failed to launch: fork/exec") &&
+		strings.Contains(processStdoutString, "operation not permitted")
+	// "operation not permitted" is handled only for the Connect My Computer case as we assume that
+	// regular SSH nodes are started as root and are unlikely to run into this error.
+	isInvalidNodePrincipal := isUsernameLookupFail || (isConnectMyComputerNode && isForkExecOperationNotPermitted)
+
+	if isInvalidNodePrincipal {
+		message := fmt.Sprintf("Invalid user. Please ensure the principal %q is a valid Linux login in the target node. Output from Node: %v",
+			sshPrincipal, processStdoutString)
+		if isConnectMyComputerNode {
+			connectMyComputerRoleName := connectmycomputer.GetRoleNameForUser(currentUser.GetName())
+			// Picking a wrong username is more probable than the role being outdated, hence why we first
+			// address the case where the actual test gets executed in step 2.
+			message = "Invalid user."
+			detailedMessage := fmt.Sprintf(
+				"If this is step 2 of the connection test, you probably picked a login which does not match "+
+					"the system user that is running Teleport Connect with the Connect My Computer agent. "+
+					"Pick the correct login and try again.\n"+
+					"If this is step 1 of the connection test and the login was picked automatically, "+
+					"it means that the role %q includes only the login %q and %q is not a valid principal for this node. "+
+					"To fix this problem, reload this page, pick Connect My Computer again, then in Teleport Connect "+
+					"remove the Connect My Computer agent and start Connect My Computer setup again.\n"+
+					"Output from the Connect My Computer agent: %v",
```

Comment:
This wall of text might be very confusing for the user
Can we somehow detect what's the actual use case?
I mean the current process's yaml config has a specific label we can check on, right?

Reply:
Does that information reach this section?
IIRC, I changed the certificate to include the `connection diagnostic id`, because this just a regular SSH session (so, no rpc calls)

Reply:
Oh, I meant that if we add another field to `TestConnectionRequest` then we won't have to fetch the role, instead we can show the appropriate message based on the new field alone.

Reply:
Then I think we should do it.

Reply:
Addressed in 606cc64.